### PR TITLE
closed #4337:label support create by font name,compatible with old labelTTF

### PR DESCRIPTION
**Pull request by WenhaiLin**

### CODE REVIEW

File: cocos/2d/CCLabel.h [hunk 1]
```diff
@@ -208,6 +239,14 @@ class CC_DLL Label : public SpriteBatchNode, public LabelProtocol
         Point position;
         Size  contentSize;
     };
+    enum class LabelType {
+
+        TypeTTF,
+        TypeBMFONT,
+        TypeCHARMAP,
+        TypeSPRITE
```
Comment (dumganhar) [inline cocos/2d/CCLabel.h]: Why not just use `TTF`, `BMFONT`, `CHARMAP` since we're using `enum class`.

File: cocos/2d/CCLabel.h [hunk 2]
```diff
@@ -243,7 +282,11 @@ class CC_DLL Label : public SpriteBatchNode, public LabelProtocol
 
     void drawShadowWithoutBlur();
 
+    void createSpriteWithFontDefinition();
+
     bool _isOpacityModifyRGB;
+    bool _contentDirty;
+    LabelType _currLabelType;
```
Comment (dumganhar) [inline cocos/2d/CCLabel.h]: _currentLabelType

File: cocos/2d/CCLabel.cpp [hunk 3]
```diff
@@ -879,10 +1027,60 @@ void Label::visit(Renderer *renderer, const kmMat4 &parentTransform, bool parent
...
+    if (fontName.find('.') != fontName.npos)
+    {
+        auto config = _fontConfig;
+        config.fontFilePath = fontName;
+        setTTFConfig(config);
+    }
+    _contentDirty = true;
+}
+
+inline const std::string& Label::getFontName() const
```
Comment (dumganhar) [inline cocos/2d/CCLabel.cpp]: Don't use `inline` in `.cpp` files.

File: cocos/2d/CCLabel.cpp [hunk 4]
```diff
@@ -50,6 +50,54 @@ Label* Label::create()
     return ret;
 }
 
+Label* Label::createWithFontDefinition(const std::string& text, FontDefinition &textDefinition)
```
Comment (dumganhar) [inline cocos/2d/CCLabel.cpp]: Why isn't the second parameter `const &`?

File: cocos/2d/CCLabel.cpp [hunk 5]
```diff
@@ -879,10 +1002,51 @@ void Label::visit(Renderer *renderer, const kmMat4 &parentTransform, bool parent
...
+        setTTFConfig(config);
+    }
+    _contentDirty = true;
+}
+
+void Label::setFontSize(int fontSize)
+{
+    if (_currentLabelType == LabelType::TTF)
+    {
+        if (_fontConfig.distanceFieldEnabled == true)
```
Comment (dumganhar) [inline cocos/2d/CCLabel.cpp]: `== true` isn't needed. If you want to check a boolean value.
Just use `if (enabled)` or `if (!enabled)`.

File: cocos/2d/CCLabel.h [hunk 6]
```diff
@@ -108,14 +121,21 @@ class CC_DLL Label : public SpriteBatchNode, public LabelProtocol
...
+     * It will create Sprite for show text if you haven't set up using TTF/BMFont/CharMap.
+     */
+    virtual void setFontDefinition(const FontDefinition& textDefinition);
+
+    /** get the text definition used by this label */
+    const FontDefinition& getFontDefinition() const { return _fontDefinition; }
+
     /** changes the string to render
-    * 
+    * @warning It is as expensive as changing the string if you havn¡¯t set up using TTF/BMFont/CharMap for label.
```
Comment (dumganhar) [inline cocos/2d/CCLabel.h]: Mess code `haven't`. What does `set up using` mean?

File: cocos/2d/CCLabel.h [hunk 7]
```diff
@@ -159,19 +179,28 @@ class CC_DLL Label : public SpriteBatchNode, public LabelProtocol
...
     virtual void setDimensions(unsigned int width,unsigned int height);
     
+    /** update content immediately.*/
+    virtual void updateContent();
+
+    virtual void setFontName(const std::string& fontName);
+    const std::string& getFontName() const { return _fontDefinition._fontName;}
+
+    virtual void setFontSize(int fontSize);
+    int getFontSize() const { return _fontDefinition._fontSize;}
```
Comment (dumganhar) [inline cocos/2d/CCLabel.h]: Why did `setFontSize` need to be `virtual`?

Comment (WenhaiLin) [inline cocos/2d/CCLabel.h]: Why not?Some people might need overwrite it in derived class.

Comment (dumganhar) [inline cocos/2d/CCLabel.h]: If some one will override the `setter`, why isn't `getter` `virtual`?

**Pull request merged**